File: app/src/main/java/com/coinninja/coinkeeper/model/dto/AddressDTO.java
package com.coinninja.coinkeeper.model.dto;

import com.coinninja.bindings.DerivationPath;
import com.coinninja.coinkeeper.model.db.Address;

import java.util.Objects;

public class AddressDTO {

    private final String address;
    private final String derivationPath;
    private final String uncompressedPublicKey;

    public static String toDerivationPathString(DerivationPath path) {
        return String.format("M/49/0/0/%s/%s", path.getChange(), path.getIndex());
    }

    @Deprecated
    public AddressDTO(Address address, String uncompressedPubKey) {
        this(address.getAddress(), toDerivationPathString(address.getDerivationPath()), uncompressedPubKey);
    }

    public AddressDTO(String address, String derivationPath, String uncompressedPublicKey) {
        this.address = address;
        this.derivationPath = derivationPath;
        this.uncompressedPublicKey = uncompressedPublicKey;
    }

    public String getDerivationPath() {
        return derivationPath;
    }

    public String getAddress() {
        return address;
    }

    public int getIndex() {
        return new DerivationPath(derivationPath).getIndex();
    }

    public String getUncompressedPublicKey() {
        return uncompressedPublicKey;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AddressDTO that = (AddressDTO) o;
        return Objects.equals(address, that.address) &&
                Objects.equals(derivationPath, that.derivationPath) &&
                Objects.equals(uncompressedPublicKey, that.uncompressedPublicKey);
    }

    @Override
    public int hashCode() {
        return Objects.hash(address, derivationPath, uncompressedPublicKey);
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/model/dto/AddressDTO.kt
package com.coinninja.coinkeeper.model.dto

import app.coinninja.cn.libbitcoin.model.DerivationPath
import com.coinninja.coinkeeper.model.db.Address
import java.util.*

class AddressDTO(val address: String, val derivationPath: String, val uncompressedPublicKey: String) {

    val index: Int get() = DerivationPath.from(derivationPath).index

    constructor(address: Address, uncompressedPubKey: String) : this(address.address, address.derivationPath.toString(), uncompressedPubKey)

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as AddressDTO?
        return address == that!!.address &&
                derivationPath == that.derivationPath &&
                uncompressedPublicKey == that.uncompressedPublicKey
    }

    override fun hashCode(): Int {
        return Objects.hash(address, derivationPath, uncompressedPublicKey)
    }

}
